### PR TITLE
Reproduce hanging build with 5.4.1

### PR DESCRIPTION
**Pull request by wolfs**
This PR tries to add a test to reproduce a hanging build.

### CODE REVIEW

**Comment by wolfs** [comment]
The PR also adds a convenient command line flag to rerun distribution tests (`--rerun`).

File: subprojects/base-services/src/main/java/org/gradle/internal/resources/DefaultResourceLockCoordinationService.java [hunk 1]
```diff
@@ -48,7 +48,7 @@ public boolean withStateLock(Transformer<ResourceLockState.Disposition, Resource
                         case RETRY:
                             resourceLockState.releaseLocks();
                             try {
-                                lock.wait();
+                                lock.wait(5000);
```
Comment (lacasseio) [inline subprojects/base-services/src/main/java/org/gradle/internal/resources/DefaultResourceLockCoordinationService.java]: Is this required to avoid test hanging on CI?

File: buildSrc/subprojects/integration-testing/src/main/kotlin/org/gradle/gradlebuild/test/integrationtests/DistributionTest.kt [hunk 2]
```diff
@@ -53,13 +55,20 @@ open class DistributionTest : Test() {
...
     init {
         dependsOn(Callable { if (binaryDistributions.distributionsRequired) listOf("all", "bin", "src").map { ":distributions:${it}Zip" } else null })
         dependsOn(Callable { if (binaryDistributions.binZipRequired) ":distributions:binZip" else null })
         dependsOn(Callable { if (libsRepository.required) ":toolingApi:publishLocalArchives" else null })
         jvmArgumentProviders.add(gradleInstallationForTest)
         jvmArgumentProviders.add(BinaryDistributionsEnvironmentProvider(binaryDistributions))
         jvmArgumentProviders.add(libsRepository)
+        outputs.upToDateWhen {
+            !rerun.getOrElse(false)
+        }
```
Comment (lacasseio) [inline buildSrc/subprojects/integration-testing/src/main/kotlin/org/gradle/gradlebuild/test/integrationtests/DistributionTest.kt]: This seems important, should we extract this into another PR and push on master?

File: subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/TaskExecutionIntegrationTest.groovy [hunk 3]
```diff
@@ -691,4 +692,190 @@ task someTask(dependsOn: [someDep, someOtherDep])
         expect:
         succeeds "custom", "--rerun-tasks"
     }
+
+    def "only must run after"() {
```
Comment (lacasseio) [inline subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/TaskExecutionIntegrationTest.groovy]: Those tests seem to be passing on master, should we include them anyway or we can drop them?

File: subprojects/core/src/main/java/org/gradle/execution/plan/Node.java [hunk 4]
```diff
@@ -38,7 +41,7 @@
 
     private ExecutionState state;
     private boolean dependenciesProcessed;
-    private boolean allDependenciesComplete;
+    private volatile boolean allDependenciesComplete;
```
Comment (lacasseio) [inline subprojects/core/src/main/java/org/gradle/execution/plan/Node.java]: Is this required outside of trying to reproduce the issue?

[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/execution/plan/Node.java]: This change is not required, this was a shot in the dark.

File: buildSrc/subprojects/integration-testing/src/main/kotlin/org/gradle/gradlebuild/test/integrationtests/DistributionTest.kt (hunk 2 above)
Comment (wolfs) [reply buildSrc/subprojects/integration-testing/src/main/kotlin/org/gradle/gradlebuild/test/integrationtests/DistributionTest.kt]: Yeah, let's put this change on `master`, so we can easily rerun the integration tests.

File: subprojects/base-services/src/main/java/org/gradle/internal/resources/DefaultResourceLockCoordinationService.java (hunk 1 above)
Comment (wolfs) [reply subprojects/base-services/src/main/java/org/gradle/internal/resources/DefaultResourceLockCoordinationService.java]: It is not. We did that to see if the lock is not notified for some reason or if the problem is something else. We don't need to timeout the lock.

@big-guy WDYT? Should we leave in the timeout?

File: subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/TaskExecutionIntegrationTest.groovy (hunk 3 above)
Comment (wolfs) [reply subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/TaskExecutionIntegrationTest.groovy]: I think this test can go. The one about the software model should stay. It passes on `master` since my changes have been reverted there.

Comment (wolfs) [reply subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/TaskExecutionIntegrationTest.groovy]: See #8715 and gradle/gradle@18b6c927b38e48eed57a5db28f9939ceb84de3d4. When #8715 is re-introduced, then the software model test will fail again. And we want to re-introduce #8715 when you are done with your changes.

Comment (lacasseio) [reply subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/TaskExecutionIntegrationTest.groovy]: I create this PR https://github.com/gradle/gradle/pull/10050 Is that enough?

Comment (wolfs) [reply subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/TaskExecutionIntegrationTest.groovy]: I think that is enough.

File: buildSrc/subprojects/integration-testing/src/main/kotlin/org/gradle/gradlebuild/test/integrationtests/DistributionTest.kt (hunk 2 above)
Comment (lacasseio) [reply buildSrc/subprojects/integration-testing/src/main/kotlin/org/gradle/gradlebuild/test/integrationtests/DistributionTest.kt]: See https://github.com/gradle/gradle/pull/10093

File: subprojects/base-services/src/main/java/org/gradle/internal/resources/DefaultResourceLockCoordinationService.java (hunk 1 above)
Comment (lacasseio) [reply subprojects/base-services/src/main/java/org/gradle/internal/resources/DefaultResourceLockCoordinationService.java]: Let's not port this then.

File: subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/TaskExecutionIntegrationTest.groovy (hunk 3 above)
Comment (lacasseio) [reply subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/TaskExecutionIntegrationTest.groovy]: The other PR prevent the creation of duplicate task at creation. All other tests should be included when re-introducing #8715.

**Comment by lacasseio** [comment]
@wolfs I will leave the additional logging inside `DefaultExecutionPlan` to you when you re-introduce #8715.
